src/providers/delimitedtext/qgsdelimitedtextprovider.cpp

```diff
@@ -604,25 +619,42 @@ void QgsDelimitedTextProvider::scanFile( bool buildIndexes )
         ( void )value.toInt( &couldBeInt[i] );
       }
 
-      if ( couldBeLongLong[i] && ! couldBeInt[i] )
+      if ( couldBeLongLong[i] && !couldBeInt[i] )
       {
         ( void )value.toLongLong( &couldBeLongLong[i] );
       }
 
-      if ( couldBeDouble[i] && ! couldBeLongLong[i] )
+      if ( couldBeDouble[i] && !couldBeLongLong[i] )
       {
         if ( ! mDecimalPoint.isEmpty() )
         {
           value.replace( mDecimalPoint, QLatin1String( "." ) );
         }
         ( void )value.toDouble( &couldBeDouble[i] );
       }
+
+      if ( couldBeDateTime[i] )
+      {
+        QDateTime dt = QDateTime::fromString( value, Qt::ISODate );
+        couldBeDateTime[i] = ( dt.isValid() && value.length() > 10 );
```

Comment:
i'd suggest moving the value.length() check to happen before the attempted conversion -- you could shortcut a bunch of work this way